### ISSUE TITLE
How to use with lua packages (e.g. citeproc-lua)

### ISSUE THREAD

**Issue by AlexChalk** [tectonic-typesetting/tectonic#1143]
I'm trying to build from a .tex file that uses [citeproc-lua](https://github.com/zepinglee/citeproc-lua). 

Tectonic is aware of the `citation-style-language.sty` package that I use in the document, but it doesn't seem to be aware of its citeproc-lua dependency, which means it doesn't understand my citation format.

Running `lualatex file.tex` with texlive (full) installed does the right thing. How can I reproduce this behaviour in tectonic?

Let me know if there's any additional info I can provide to help with this.

**Comment by AlexChalk** [comment]
n.b. there's a second README.md with some additional info in the citeproc-lua project: https://github.com/zepinglee/citeproc-lua/blob/ab3ce712cc92073f12be26ff0b22b30eb906092d/examples/README.md

It includes the following:

> For LuaLaTeX, the citations and bibliography can be generated without triggering BibTeX. It takes at most two passes to get the correct labels.

```bash
cd example
lualatex example.tex
lualatex example.tex
```

> The `latexmk` can also be used.
```bash
latexmk -cd -lualatex examples/example.tex
```

> For engines other than LuaLaTeX, the `citeproc` executable is required to run as BibTeX.

```bash
cd examples
pdflatex example.tex
citeproc-lua example.aux
pdflatex example.tex
```

Is there any info here I could use to get the package working with tectonic?

**Comment by rm-dr** [comment]
Tectonic does not use lualatex, so citeproc-lua won't work as you expect.
You could try the "for engines other than lualatex" process.

**Comment by AlexChalk** [comment]
Hi @rm-dr, 

Is there any interest in adding this feature? I don't know what lualatex is doing under the hood (I'm new to latex), but even if it's not reproducible in tectonic, the "other libraries" behaviour described above seems like it could be specified and automated?

I'd be happy to help with this.

**Comment by rm-dr** [comment]
LuaLaTeX allows you to run Lua inside your document. No other compiler (including Tectonic) supports this. Some packages (like citeproc) require lua, and won't work without external tools.

Lua scripts would be pretty hard to integrate into tectonic. Tectonic is built on XeTeX, and I expect we'd need a complete rewrite to use lualatex instead.

**Comment by AlexChalk** [comment]
So my understanding of build tools like tectonic and xetex is that they do multiple passes when generating pdfs. Couldn't you spec how tectonic does those passes, e.g. some json that communicates something like (pseudocode)
```
if uses_citeproc_lua_package:
  run first pass
  run citeproc-lua on .aux file
  run next pass
```
It seems like that lets you incorporate other ecosystem packages into the build process without rewriting the compiler.

**Comment by pkgw** [comment]
@AlexChalk A big point of the Lua integrations is that there's Lua code that can run within the TeX engine itself, during your "first pass" / "second pass". I'm not familiar with this particular package, but I have to imagine that's part of how it's implemented. And, for better or for worse, the LuaLaTeX and XeTeX engine implementations (the latter being what Tectonic is based on) have diverged sufficiently that I would be extremely skeptical that you could integrate compatible Lua support into Tectonic's codebase.

As a first approximation: if it involves Lua, it won't work with Tectonic, unfortunately.

**Comment by AlexChalk** [comment]
There's a workaround that doesn't require lualatex though:
```
cd examples
pdflatex example.tex
citeproc-lua example.aux
pdflatex example.tex
```
This behaviour couldn't be integrated into tectonic? You would need to pull down the lua package (and lua too), and you'd need a way of telling tectonic what to do between runs, but I'm confused by why that would involve an engine implementation change.

**Comment by rm-dr** [comment]
You're right---the workaround wouldn't require us to re-design the engine.

In theory, it's possible to have tectonic call external tools, but I'm not sure how well such features would fit into Tectonic's "feature model." Tectonic isn't a build tool (like `latexmk`), it's a latex engine (like `pdflatex`).

One idea is to add a mechanism to call extra tools in `Tectonic.toml`, but that may or may not be a good solution. One of the goals of Tectonic is strong reproducibility, and calling external tools contradicts this: we can't guarantee what version of a tool we're using, for example, unless we compile that tool into tectonic (like we do with bibtex).

**Comment by AlexChalk** [comment]
No worries, and I now see how adding everything needed for citeproc-lua to tectonic's backing bundle would be a meaningful feature change. I also found https://github.com/tectonic-typesetting/tectonic/issues/893 which seems like a somewhat analogous issue.

As a user I'd love to see it—I prefer the UX of tectonic to downloading all of texlive for the 0.1% of it I use, but if tectonic doesn't pull everything I need, I end up downloading all of texlive anyway 😅. As a dev, I get that we have jobs and limited time, and that you'd probably need a build tool like nix to do this in a reproducible way, which is a lot. 

I'll close this for now, as I think the issue I linked already captures the wider design/featureset discussion.
